lib/src/eol.rs

```diff
@@ -0,0 +1,339 @@
+    }
+}
+
+fn is_binary(bytes: &[u8]) -> bool {
+    // TODO(06393993): align the algorithm with git so that the git config autocrlf
+    // users won't see different decisions on whether a file is binary and needs to
+    // perform EOL conversion.
+    bytes.contains(&b'\0')
+}
+
+pub(crate) struct TargetEolStrategy {
+    eol_conversion_settings: EolConversionSettings,
+}
+
+impl TargetEolStrategy {
+    /// The limit is to probe whether the file is binary is 8KB.
+    const PROBE_LIMIT: u64 = 8 << 10;
+
+    pub(crate) async fn convert_eol_for_snapshot<'a>(
+        &self,
+        mut contents: impl AsyncRead + Send + Unpin + 'a,
+    ) -> Result<impl AsyncRead + Send + Unpin + 'a, std::io::Error> {
+        match self.eol_conversion_settings {
+            EolConversionSettings::None => {
+                Ok(Box::pin(contents) as Pin<Box<dyn AsyncRead + Send + Unpin>>)
+            }
+            EolConversionSettings::Input | EolConversionSettings::InputOutput => {
+                let mut peek = vec![];
+                (&mut contents)
+                    .take(Self::PROBE_LIMIT)
+                    .read_to_end(&mut peek)
+                    .await?;
+                let target_eol = if is_binary(&peek) {
+                    TargetEol::PassThrough
+                } else {
+                    TargetEol::Lf
+                };
+                let peek = Cursor::new(peek);
+                let contents = peek.chain(contents);
+                convert_eol(contents, target_eol).await
+            }
+        }
+    }
+
+    pub(crate) async fn convert_eol_for_update<'a>(
+        &self,
+        mut contents: impl AsyncRead + Send + Unpin + 'a,
+    ) -> Result<impl AsyncRead + Send + Unpin + 'a, std::io::Error> {
+        match self.eol_conversion_settings {
+            EolConversionSettings::None | EolConversionSettings::Input => {
+                Ok(Box::pin(contents) as Pin<Box<dyn AsyncRead + Send + Unpin>>)
+            }
+            EolConversionSettings::InputOutput => {
+                let mut peek = vec![];
+                (&mut contents)
+                    .take(Self::PROBE_LIMIT)
+                    .read_to_end(&mut peek)
+                    .await?;
+                let target_eol = if is_binary(&peek) {
+                    TargetEol::PassThrough
+                } else {
+                    TargetEol::Crlf
+                };
+                let peek = Cursor::new(peek);
+                let contents = peek.chain(contents);
+                convert_eol(contents, target_eol).await
+            }
+        }
+    }
+}
+
+/// Configuring auto-converting CRLF line endings into LF when you add a file to
+/// the backend, and vice versa when it checks out code onto your filesystem.
+#[derive(Debug, PartialEq, Eq, Copy, Clone, serde::Deserialize)]
+#[serde(rename_all(deserialize = "kebab-case"))]
+#[non_exhaustive]
+pub enum EolConversionSettings {
```

Comment:
nit: `Settings` sounds like a struct with multiple fields. Maybe rename to `EolConversion`, `EolConversionMode`, etc.?